### PR TITLE
[datasets] Expand `to_torch`'s functionality

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Expands the `to_torch` method for Datasets with:
* An ability to choose to output a list/dict of feature tensors instead of just one (through setting `feature_columns` to be a list of lists or a dict of lists)
* An ability to choose whether the label should be unsqueezed or not
* An ability to pass `None` as the label (for prediction).

Furthermore, this changes how the `feature_column_dtypes` argument works. Previously, it took a list of dtypes for each feature. However, as the tensor was concatenated in the end, only one dtype mattered (the biggest one). Now, this argument expects a single dtype which will be applied to the features tensor (or a list/dict if `feature_columns` is a list of list/dict of lists).

Unit tests for all cases are included.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Docstrings don't render properly, btw
Screen Shot 2021-12-15 at 3 00 29 PM

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -1919,29 +1924,55 @@ def to_torch(self,
...
-        ``batch_size`` used by the DataLoader, and n is the number of features.
+        elements. The first item are the features, and the second item
+        is the label. Those can take on different forms, dependning on
+        the specified arguments.
+
+        For the features tensor (N is the ``batch_size`` used by the
+        DataLoader,  and n, m, k are the number of features per tensor):
+
+        * If ``feature_columns`` is a ``List[str]``, the features will be
+          a tensor of shape (N, n), with columns corresponding to =
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:1935]:
```suggestion
          a tensor of shape (N, n), with columns corresponding to
```

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -1919,29 +1924,55 @@ def to_torch(self,
...
+        * If ``feature_columns`` is a ``Dict[str, List[str]]``, the features
+          will be a dict of key-tensor pairs of shape
+          {key1: (N, m),..., keyN: (N, k)}, with columns of each
+          tensor corresponding to the value of ``feature_columns`` under the
+          key.
+
+        If ``unsqueeze_label_tensor=True`` (default), the label tensor will be
+        of shape (N, 1). Otherwise, it will be of shape (N,). It can also be
+        specifed as None for prediction (note that it will be still present
+        in the tuple yielded by the generator).
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:1951]: > It can also be
        specifed as None for prediction (note that it will be still present
        in the tuple yielded by the generator).

Wording of this is a little weird - I thought this was referring to `unsqueeze_label_tensor`, not `label_column`.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -1919,29 +1924,55 @@ def to_torch(self,
...
-                (second element of the output list).
-            feature_columns (Optional[List[str]]): The names of the columns
-                to use as the features. If None, then use all columns
-                except the label columns as the features.
+            label_column (Optional[str]): The name of the column used as the
+                label (second element of the output list). Can be None for
+                prediction, in which case the second element of returned
+                tuple will also be None.
+            feature_columns (Union[None, List[str], List[List[str]],
+                Dict[str, List[str]]]): The names of the columns
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:1964]: Generated docs are weird here... not sure what's going on.

image

File: python/ray/data/tests/test_dataset.py [hunk 4]
```diff
@@ -2400,28 +2400,101 @@ def test_to_torch(ray_start_regular_shared, pipelined):
...
+                assert all(item.dtype == torch.long for item in features)
+            features = torch.cat(tuple(features), dim=1)
+        elif input == "dict":
+            assert isinstance(features, dict)
+            assert all(
+                isinstance(item, torch.Tensor) for item in features.values())
+            if force_dtype:
+                assert all(
+                    item.dtype == torch.long for item in features.values())
+            features = torch.cat(tuple(features.values()), dim=1)
```
Comment (matthewdeng) [inline python/ray/data/tests/test_dataset.py:2479]: Is order guaranteed here?

File: python/ray/data/tests/test_dataset.py [hunk 5]
```diff
@@ -2400,28 +2400,101 @@ def test_to_torch(ray_start_regular_shared, pipelined):
...
+            assert isinstance(label, torch.Tensor)
+            if force_dtype:
+                assert label.dtype == torch.long
+            if unsqueeze_label_tensor:
+                assert label.dim() == 2
+            else:
+                assert label.dim() == 1
+                label = label.view(-1, 1)
+            features = torch.cat((features, label), dim=1)
+        iterations.append(features.numpy())
```
Comment (matthewdeng) [inline python/ray/data/tests/test_dataset.py:2493]: nit: Is there some better term for features + label combined? Attributes?

**Review by matthewdeng — COMMENTED** [review]
This is really great for improving usability, thanks for adding this!

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -1959,48 +1995,70 @@ def to_torch(self,
...
-        if feature_columns and feature_column_dtypes:
-            if len(feature_columns) != len(feature_column_dtypes):
-                raise ValueError("The lengths of `feature_columns` "
-                                 f"({len(feature_columns)}) and "
-                                 f"`feature_column_dtypes` ("
-                                 f"{len(feature_column_dtypes)}) do not "
-                                 "match!")
+        multi_input = feature_columns and (isinstance(feature_columns, dict) or
+                                           isinstance(feature_columns[0],
+                                                      (list, tuple)))
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/data/dataset.py:2000]: While it's not in the type hint, passing a tuple should also be allowed as in this case it's identical to a list

File: python/ray/data/dataset.py [hunk 7]
```diff
@@ -1959,48 +1995,70 @@ def to_torch(self,
...
+                            for idx, _ in enumerate(feature_columns)
+                        ]
                 yield (features_tensor, label_tensor)
 
         return TorchIterableDataset(make_generator)
 
     def to_tf(self,
               *,
-              label_column: str,
+              label_column: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/data/dataset.py:2061]: This shouldn't be there, good catch

File: python/ray/data/tests/test_dataset.py (hunk 4 above)
Comment (Yard1) [reply python/ray/data/tests/test_dataset.py:2479]: Yes

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -1959,48 +1995,70 @@ def to_torch(self,
...
+                    features_tensor = get_feature_tensors(
+                        batch, feature_columns, feature_column_dtypes)
+                else:
+                    if isinstance(feature_columns, dict):
+                        features_tensor = {
+                            key: get_feature_tensors(
+                                batch, feature_columns[key],
+                                feature_column_dtypes[key]
+                                if isinstance(feature_column_dtypes,
+                                              dict) else feature_column_dtypes)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/data/dataset.py]: Added some validation

File: python/ray/data/dataset.py (hunk 3 above)
Comment (Yard1) [reply python/ray/data/dataset.py]: Fixed

File: python/ray/data/dataset.py [hunk 9]
```diff
@@ -1959,48 +1995,70 @@ def to_torch(self,
...
-                    batch = batch[feature_columns]
-
-                if feature_column_dtypes:
-                    dtypes = feature_column_dtypes
+                if label_column:
+                    label_vals = batch.pop(label_column).values
+                    label_tensor = torch.as_tensor(
+                        label_vals, dtype=label_column_dtype)
+                    if unsqueeze_label_tensor:
+                        label_tensor = label_tensor.view(-1, 1)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/data/dataset.py:2042]: I will leave it as view(-1, 1) just in case unsqueeze has some side effects

File: python/ray/data/dataset.py [hunk 10]
```diff
@@ -1961,41 +1998,99 @@ def to_torch(self,
...
+                        "`feature_column_dtypes` must be None, `torch.dtype`,"
+                        f" or dict, got {type(feature_column_dtypes)}.")
+                if set(feature_columns) != set(feature_column_dtypes):
+                    raise ValueError(
+                        "`feature_columns` and `feature_column_dtypes` "
+                        "must have the same keys.")
+            elif isinstance(feature_columns[0], (list, tuple)):
+                if not isinstance(feature_column_dtypes, (list, tuple)):
+                    raise TypeError(
+                        "If `feature_columns` is a dict, "
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:2023]:
```suggestion
                        "If `feature_columns` is a list of lists, "
```

File: python/ray/data/dataset.py [hunk 11]
```diff
@@ -1921,29 +1926,56 @@ def to_torch(self,
...
+        * If ``feature_columns`` is a ``Dict[str, List[str]]``, the features
+          will be a dict of key-tensor pairs of shape
+          {key1: (N, m),..., keyN: (N, k)}, with columns of each
+          tensor corresponding to the value of ``feature_columns`` under the
+          key.
+
+        If ``unsqueeze_label_tensor=True`` (default), the label tensor will be
+        of shape (N, 1). Otherwise, it will be of shape (N,). ``label_column``
+        can also be specifed as None for prediction (note that it will be
+        still present in the tuple yielded by the generator).
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:1953]: This is a bit confusing to me. I guess the underlying logic is that the tensor will be included in the output table as one of the feature columns, but even that is only if `feature_columns` is not specified.

Would something like this be more clear?
> If ``label_column`` is specified as ``None``, then no column from the ``Dataset`` will be treated as the label, and the  output label tensor will be ``None``.

File: python/ray/data/dataset.py [hunk 12]
```diff
@@ -1961,41 +1998,99 @@ def to_torch(self,
...
-                                 f"({len(feature_columns)}) and "
-                                 f"`feature_column_dtypes` ("
-                                 f"{len(feature_column_dtypes)}) do not "
-                                 "match!")
+        multi_input = feature_columns and (isinstance(feature_columns, dict) or
+                                           isinstance(feature_columns[0],
+                                                      (list, tuple)))
+
+        if not feature_columns:
+            feature_columns = None
```
Comment (matthewdeng) [inline python/ray/data/dataset.py:2006]: nit: Would prefer to add a comment here describing this implicit behavior, or just raising an error here requiring the user to explicitly choose between `None` and  non-empty columns.

**Review by matthewdeng — APPROVED** [review]

File: python/ray/data/dataset.py (hunk 11 above)
Comment (matthewdeng) [reply python/ray/data/dataset.py:1953]:
```suggestion
        of shape (N, 1). Otherwise, it will be of shape (N,). 
        If ``label_column`` is specified as ``None``, then no column from the 
        ``Dataset`` will be treated as the label, and the output label tensor 
        will be ``None``.
```

File: python/ray/data/dataset.py (hunk 12 above)
Comment (matthewdeng) [reply python/ray/data/dataset.py:2006]:
```suggestion
       # If an empty collection is passed in, treat it the same as None
        if not feature_columns:
            feature_columns = None
```

File: python/ray/data/dataset.py [hunk 13]
```diff
@@ -1952,6 +1985,11 @@ def to_torch(self,
                 if the dataset size is not divisible by the batch size. If
                 False and the size of dataset is not divisible by the batch
                 size, then the last batch will be smaller. Defaults to False.
+            unsqueeze_label_tensor (bool): If set to True, the label tensor
```
Comment (scv119) [inline python/ray/data/dataset.py:1988]: will name it `squeeze_label_tensor` and default to False sound better?

File: python/ray/data/tests/test_dataset.py [hunk 14]
```diff
@@ -2400,28 +2400,102 @@ def test_to_torch(ray_start_regular_shared, pipelined):
         assert np.array_equal(np.sort(df.values), np.sort(combined_iterations))
 
 
-def test_to_torch_feature_columns(ray_start_regular_shared):
+@pytest.mark.parametrize("input", ["single", "list", "dict"])
+@pytest.mark.parametrize("force_dtype", [False, True])
+@pytest.mark.parametrize("label_type", [None, "squeezed", "unsqueezed"])
+def test_to_torch_feature_columns(ray_start_regular_shared, input, force_dtype,
+                                  label_type):
```
Comment (scv119) [inline python/ray/data/tests/test_dataset.py:2407]: does this test squeeze labels?

Comment (Yard1) [reply python/ray/data/tests/test_dataset.py:2407]: It does

File: python/ray/data/dataset.py (hunk 13 above)
Comment (Yard1) [reply python/ray/data/dataset.py:1988]: Perhaps, though the operation being performed is unsqueezing, as the label tensors are created squeezed by default (by torch). I feel like `unsqueeze_label_tensor` better conveys the fact that an additional operation is carried out

**Review by scv119 — APPROVED** [review]

File: python/ray/data/dataset.py [hunk 15]
```diff
@@ -1921,29 +1926,57 @@ def to_torch(self,
...
+        * If ``feature_columns`` is a ``Dict[str, List[str]]``, the features
+          will be a dict of key-tensor pairs of shape
+          {key1: (N, m),..., keyN: (N, k)}, with columns of each
+          tensor corresponding to the value of ``feature_columns`` under the
+          key.
+
+        If ``unsqueeze_label_tensor=True`` (default), the label tensor will be
+        of shape (N, 1). Otherwise, it will be of shape (N,). 
+        If ``label_column`` is specified as ``None``, then no column from the 
+        ``Dataset`` will be treated as the label, and the output label tensor
```
Comment (Yard1) [inline python/ray/data/dataset.py:1953]:
```suggestion
        of shape (N, 1). Otherwise, it will be of shape (N,).
        If ``label_column`` is specified as ``None``, then no column from the
        ``Dataset`` will be treated as the label, and the output label tensor
```
